File: src/main/java/net/minestom/server/utils/debug/DebugUtils.java
package net.minestom.server.utils.debug;

import net.minestom.server.utils.PropertyUtils;
import org.jetbrains.annotations.ApiStatus;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Utils class useful for debugging purpose.
 */
@ApiStatus.Internal
public final class DebugUtils {
    public static boolean INSIDE_TEST = PropertyUtils.getBoolean("minestom.inside-test", false);

    private final static Logger LOGGER = LoggerFactory.getLogger(DebugUtils.class);

    private static final String LINE_SEPARATOR = System.getProperty("line.separator");

    /**
     * Prints the current thread stack trace elements.
     */
    public static synchronized void printStackTrace() {
        StackTraceElement[] elements = Thread.currentThread().getStackTrace();

        StringBuilder stringBuilder = new StringBuilder();

        stringBuilder.append("START STACKTRACE");
        stringBuilder.append(LINE_SEPARATOR);

        for (int i = 0; i < Integer.MAX_VALUE; i++) {
            if (i >= elements.length)
                break;

            final StackTraceElement element = elements[i];
            final String line = element.getClassName() + "." + element.getMethodName() + " (line:" + element.getLineNumber() + ")";
            stringBuilder.append(line);
            stringBuilder.append(LINE_SEPARATOR);
        }

        stringBuilder.append("END STACKTRACE");

        LOGGER.info(stringBuilder.toString());
    }

}

File: src/main/java/net/minestom/server/utils/debug/DebugUtils.kt
package net.minestom.server.utils.debug

import net.minestom.server.utils.PropertyUtils
import java.lang.StackTraceElement
import java.lang.StringBuilder
import net.minestom.server.utils.debug.DebugUtils
import org.jetbrains.annotations.ApiStatus
import org.slf4j.LoggerFactory

/**
 * Utils class useful for debugging purpose.
 */
@ApiStatus.Internal
object DebugUtils {
    @JvmField
    var INSIDE_TEST = PropertyUtils.getBoolean("minestom.inside-test", false)
    val LOGGER = LoggerFactory.getLogger(DebugUtils::class.java)
    private val LINE_SEPARATOR = System.getProperty("line.separator")

    /**
     * Prints the current thread stack trace elements.
     */
    @Synchronized
    fun printStackTrace() {
        val elements = Thread.currentThread().stackTrace
        val stringBuilder = StringBuilder()
        stringBuilder.append("START STACKTRACE")
        stringBuilder.append(LINE_SEPARATOR)
        for (i in 0 until Int.MAX_VALUE) {
            if (i >= elements.size) break
            val element = elements[i]
            val line = element.className + "." + element.methodName + " (line:" + element.lineNumber + ")"
            stringBuilder.append(line)
            stringBuilder.append(LINE_SEPARATOR)
        }
        stringBuilder.append("END STACKTRACE")
        LOGGER.info(stringBuilder.toString())
    }
}